### ISSUE TITLE
Bug (type-utils): `URL` cannot be referenced from `@types/node` in a `TypeOrValueSpecifier`

### ISSUE THREAD

**Issue by abrahamguo** [typescript-eslint/typescript-eslint#9608]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

I expect that `URL` should be able to be referenced from a `{ from: 'package', name: 'URL', package: 'node' }` `TypeOrValueSpecifier`, but it is not.

### Reproduction Repository Link

https://github.com/abrahamguo/repro/tree/URL

### Repro Steps

1) clone the repo
2) `npm i`
3) `npx tsc`
4) `node .`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `7.16.1` or `8.0.0-alpha.49` |
| `@typescript-eslint/parser`            | `7.16.1` or `8.0.0-alpha.49` |
| `@typescript-eslint/scope-manager`     | `7.16.1` or `8.0.0-alpha.49` |
| `@typescript-eslint/typescript-estree` | `7.16.1` or `8.0.0-alpha.49` |
| `@typescript-eslint/type-utils`        | `7.16.1` or `8.0.0-alpha.49` |
| `@typescript-eslint/utils`             | `7.16.1` or `8.0.0-alpha.49` |
| `TypeScript`                           | `5.5.3` |
| `ESLint`                               | `9.7.0` |
| `node`                                 | `21.6.1` |

**Comment by JoshuaKGoldberg** [comment]
Whoo, yeah, this is tricky.

**tl;dr:**: `URL` _can_ be referenced _if_ you import it from `url` & use that in the package specifier. It can't be referenced if you either:

* Import from `node:url` _(the recommended path to import from!)_
* Use the global `URL` without importing

---

First, `node_modules/@types/node/url.d.ts` is set up like this:

```ts
declare module "url" {
  // ... (actual declarations here) ...
}
declare module "node:url" {
    export * from "url";
}
```

...which means our existing _"what **single** module name is this exported under?"_ logic might need to be expanded to also support re-exports.

Second, the declaration of `url`/`node:url`'s `URL` looks like this (trimming out some stuff):

```ts
declare module "url" {
    class URL { /* ... */ }

    import { URL as _URL, URLSearchParams as _URLSearchParams } from "url";

    global {
        interface URL extends _URL {}
        interface Global {
            URL: typeof _URL;
        }
        /** `URL` class is a global reference for `require('url').URL` */
        var URL: typeof globalThis extends {
            onmessage: any;
            URL: infer T;
        } ? T
            : typeof _URL;
    }
}
```

Note that there's a difference between:

* The _class_: exported from `url`/`node:url`
* The _interface_: declared to be a global

Right now the linked repro refers to the global URL... which uses the `typeof globalThis extends { ... infer T } ? T : typeof _URL` shenanigan to refer to the global `interface`.

**Comment by JoshuaKGoldberg** [comment]
Whoo, yeah, this is tricky.

**tl;dr:** `URL` _can_ be referenced _if_ you import it from `url` & use that in the package specifier. It can't be referenced if you either:

* Import from `node:url` _(the recommended path to import from!)_
* Use the global `URL` without importing

Accepting PRs!

---

First, `node_modules/@types/node/url.d.ts` is set up like this:

```ts
declare module "url" {
  // ... (actual declarations here) ...
}
declare module "node:url" {
    export * from "url";
}
```

...which means our existing _"what **single** module name is this exported under?"_ logic might need to be expanded to also support re-exports.

Second, the declaration of `url`/`node:url`'s `URL` looks like this (trimming out some stuff):

```ts
declare module "url" {
    class URL { /* ... */ }

    import { URL as _URL, URLSearchParams as _URLSearchParams } from "url";

    global {
        interface URL extends _URL {}
        interface Global {
            URL: typeof _URL;
        }
        /** `URL` class is a global reference for `require('url').URL` */
        var URL: typeof globalThis extends {
            onmessage: any;
            URL: infer T;
        } ? T
            : typeof _URL;
    }
}
```

Note that there's a difference between:

* The _class_: exported from `url`/`node:url`
* The _interface_: declared to be a global

Right now the linked repro refers to the global URL... which uses the `typeof globalThis extends { ... infer T } ? T : typeof _URL` shenanigan to refer to the global `interface`.

**Comment by abrahamguo** [comment]
Thanks for the investigation!

To clarify a bit more, this whole thing appears to only be a problem if `lib` is set to exclude the `URL` from `lib: dom`.
